genesis/sensors/base_sensor.py

```diff
@@ -12,10 +21,127 @@ class Sensor(RBC):
+        self._options: "SensorOptions" = sensor_options
+        self._idx: int = sensor_idx
+        self._manager: "SensorManager" = sensor_manager
+        self._cache_idx: int = -1  # cache_idx is set by the SensorManager during the scene build phase
+        self._read_delay_steps: int = 0
+        self._cache_buffer_length: int = 0
+
+    # =============================== implementable methods ===============================
+
+    @gs.assert_unbuilt
+    def build(self):
+        """
+        This method is called by the SensorManager during the scene build phase to initialize the sensor.
+        This is where any shared metadata should be initialized.
+        """
+        delay_steps_float = self._options.read_delay / self._manager._sim.dt
+        self._read_delay_steps = round(delay_steps_float)
+        if not np.isclose(delay_steps_float, self._read_delay_steps):
+            gs.logger.warn(
+                f"Read delay should be a multiple of the simulation time step. Got {self._options.read_delay} and "
+                f"{self._manager._sim.dt}. Actual read delay will be {1/self._read_delay_steps}."
+            )
+        self._cache_buffer_length = self._read_delay_steps + 1
+
+    def _get_return_format(self) -> dict[str, tuple[int, int]] | tuple[int, int] | None:
+        """
+        Data format of the read() return value.
+        dict: a dictionary with string keys and tensor values will be returned.
+              The tuple (start_idx, end_idx) will be used to get a slice of the values from the cache.
+        tuple: the slice of the cache will be returned.
+        None: the entire tensor (B, cache_length, cache_shape) will be returned.
+        """
+        raise NotImplementedError("Sensors must implement `return_format()`.")
+
+    @gs.assert_built
+    def _update_shared_gt_cache(self):
+        """
+        Update the shared sensor ground truth cache for all sensors of this class.
+        Shared information is stored in SensorManager
+        """
+        raise NotImplementedError("Sensors must implement `update_shared_gt_cache()`.")
+
     @gs.assert_built
-    def read(self, envs_idx: Optional[List[int]] = None):
+    def _update_shared_cache(self):
         """
-        Read the sensor data.
-        Sensor implementations should ideally cache the data to avoid unnecessary computations.
+        Update the shared sensor cache for all sensors of this class using information stored in shared metadata.
+        """
+        raise NotImplementedError("Sensors must implement `update_shared_cache()`.")
+
+    def _get_cache_length(self) -> int:
+        """
+        The length (first dimension of cache shape) of the cache for this sensor instance.
+        The sum of cache_length for all sensors of this type determines the length of the cache.
+        """
+        raise NotImplementedError("Sensors must implement `cache_length()`.")
+
+    @classmethod
+    def _get_cache_size(cls) -> int:
```

Comment:
Having 'get_cache_size', 'get_cache_length', and 'get_return_format' is too much. Let's discuss this together.